subsys/net/ip/tcp.c

```diff
@@ -2042,16 +2050,24 @@ NET_CONN_CB(tcp_established)
 
 	net_context_set_appdata_values(pkt, IPPROTO_TCP);
 
-	data_len = net_pkt_appdatalen(pkt);
-	if (data_len > net_tcp_get_recv_wnd(context->tcp)) {
-		NET_ERR("Context %p: overflow of recv window (%d vs %d), "
-			"pkt dropped",
-			context, net_tcp_get_recv_wnd(context->tcp), data_len);
-		return NET_DROP;
+	if (!context->tcp->zero_wnd_sent) {
+		data_len = net_pkt_appdatalen(pkt);
+		if (data_len > net_tcp_get_recv_wnd(context->tcp)) {
+			NET_ERR("Context %p: overflow of recv window "
+				"(%d vs %d), pkt dropped",
+				context, net_tcp_get_recv_wnd(context->tcp),
+				data_len);
+			return NET_DROP;
+		}
+	} else {
+		/* Zero window probe expected. */
```

Comment:
This comment is confusing: nothing is expected, we just do treat the current segment as zero window probe.